src/cargo/core/compiler/mod.rs

```diff
@@ -235,13 +235,23 @@ fn rustc(cx: &mut Context<'_, '_>, unit: &Unit, exec: &Arc<dyn Executor>) -> Car
     } else {
         format!("{}.d", unit.target.crate_name())
     };
+
     let rustc_dep_info_loc = root.join(dep_info_name);
     let dep_info_loc = fingerprint::dep_info_loc(cx, unit);
 
     rustc.args(cx.bcx.rustflags_args(unit));
     if cx.bcx.config.cli_unstable().binary_dep_depinfo {
         rustc.arg("-Z").arg("binary-dep-depinfo");
     }
+
+    if unit.target.get_binary_name().is_some() {
+        let mut filepath = String::from("--emit=link=");
+        for output in outputs.iter() {
+            filepath.push_str(&output.path.to_string_lossy());
+        }
```

Comment:
This doesn't seem to be the correct way to get the link filename. For example, on macOS this will say `--emit=link=/Users/eric/foo/target/debug/deps/z62/Users/eric/foo/target/debug/deps/z62.dSYM`   Notice the path is there twice, and it is using the wrong filename.  It might be worth investigating why the output paths are not being computed correctly.

I think this would need to call `find()` to find the output with a `FileFlavor::Normal`.

Also, I would think this `--emit` logic would be down in `build_base_args` where the `--emit` statements are. Is there a reason it is up here?

Reply:
> Also, I would think this `--emit` logic would be down in `build_base_args` where the `--emit` statements are. Is there a reason it is up here?

The `build_base_args` is computed much earlier than we compute the `outputs` variable.

Reply:
>  Notice the path is there twice, and it is using the wrong filename. It might be worth investigating why the output paths are not being computed correctly.

I think this issue will fix itself, if we select the right element from the `outputs` vector. I will try implementing the `find()` logic you have suggested. If that doesn't work then we will try to figure out a way to debug and investigate this.